### PR TITLE
BIP-0047: Reusable payment codes

### PR DESCRIPTION
**Pull request by justusranvier**
Payment codes are SPV-friendly alternatives to DarkWallet-style stealth addresses which provide useful features such as positively identifying senders to recipients and automatically providing for transaction refunds.

Payment codes can be publicly advertised and associated with a real-life identity without causing a loss of financial privacy.

Compared to stealth addresses, payment codes require less blockchain data storage.

Payment codes require 65 bytes of OP_RETURN data per sender-recipient pair, while stealth addresses require 40 bytes per transaction.

### CODE REVIEW

File: bip-0047.mediawiki
Comment (dskloet) [inline bip-0047.mediawiki]: There's non mention of a version byte in the previous section. But the first byte (type) is described as 0x01, not 0x37.

File: bip-0047.mediawiki
Comment (dskloet) [inline bip-0047.mediawiki]: I don't see what this has to do with ASCII.

File: bip-0047.mediawiki
Comment (dskloet) [inline bip-0047.mediawiki]: Remove "a" or make "codes" singular.

File: bip-0047.mediawiki
Comment (dskloet) [inline bip-0047.mediawiki]: Since Alice got Bob's payment code out of band, I wonder if it makes sense to allow Alice to send her payment code to Bob out of band as well. That way you can avoid worrying about leaking the notification address.

File: bip-0047.mediawiki
Comment (justusranvier) [inline bip-0047.mediawiki]: The ASCII reference is a left over from before we were using a BIP number for that path level.

File: bip-0047.mediawiki
Comment (justusranvier) [inline bip-0047.mediawiki]: Alice getting Bob's payment code out of band might mean that it was printed on a business card which Alice got from Bob at a conference, or on paper fliers, or embedded into a DNS TXT record for a web site.

We're trying to avoid requirements for interactive communication here, in particular it would make payment codes vastly less useful if the recipient had to be online in order to receive payments.

The reason that Alice sends her payment code to Bob via a notification transaction, even if she also uses some other communication channel at the same time, is to preserve Bob's ability to recover his wallet from a seed.

**Comment by laanwj** [comment]
Ready for merging?

**Comment by justusranvier** [comment]
There's a correction to the serialization format that I want to push before this will be ready for merge.

File: bip-0047.mediawiki
Comment (dskloet) [inline bip-0047.mediawiki]: Except for charity donations I can't think of any use cases where you'd want to send money without prior communication. Maybe you can include some examples in the text?

Being able to recover the wallet from a seed is a good point. Though the protocol could be changed to make Bob responsible for recording the sender payment codes in the blockchain. I'm not saying one is better than the other but I think it would be good to allow for both options. It especially worries me that the send has to take special care not to mix up the coins in the notification address.

In any case it would be good to mention the argument about being able to recover the wallet from a seed in the text.

**Comment by erasmospunk** [comment]
How do the payment codes interact with the URI scheme (BIP21) and payment protocol (BIP70)?

For the URI, there could be a new parameter `paycode`:

```
bitcoin:1FpVXFZ3ZYE6SLg8T2U1pTKgySx46rmq2Z?amount=0.01&paycode=PXXXXXXXXXX
```

Like this, incompatible clients will gracefully degrade. In case of the BIP72 `r=` parameter, it would take precedence as the spec dictates.

There could be separate BIPs that define this behavior, but it would be nice to have everything in one place.

**Comment by justusranvier** [comment]
If it would be appropriate, I could extend this pull request to update BIP 21 and/or BIP 70 as well.

**Comment by erasmospunk** [comment]
When we are using BIP-70 the client always gets a set of addresses (outputs) to pay, so paycodes seem redundant from a privacy perspective.

On the other hand one scenarios that comes in mind:

Alice joins an online dice game, she scans the qr-code and connects with the BIP-70 compatible server that sends her a paycode. She then replies to the server with the BIP-47 notification and the first bet transaction. This creates the dice game paycode/identity in her wallet, that she can create more transactions without talking again to the BIP-70 server. So after she won, she sends again to the dice paycode/identity (without sending again the notification) and retains her privacy. This emulates the famous `1diceXXXX` address but with the additional benefit of the privacy.

Another usage is in an exchange, where you use paycodes to fill the account balance and using different address in subsequent deposits. The wallet could even send the refund address as a paycode so that the  withdrawals are also privacy protected.

In BIP-47 we need a BIP-70 notification flag additionally to the Bitmessage notification.
In BIP-70 an extension needs to be created to support the above use cases. Probably adding an `optional bytes paycode` to the `message Output`.

**Comment by justusranvier** [comment]
The BIP-47 workflow intentionally differs from BIP-70 in that it's designed for the payer to choose the outputs instead of the payee. This gives the sender more control over their privacy, because they have the option to use multiple blockchain transactions to complete a single business transaction.

It's also designed for a TOFU security model, where only the initial payment code exchange needs to be verified instead of every subsequent payment address.

I'm not sure how easy it would be to make that workflow fit into BIP-70.

**Comment by kristovatlas** [comment]
@justusranvier is this ready to merge?

**Comment by justusranvier** [comment]
As far as I know there is no barrier to merging it in its current form, other than lacking yet-to-be-developed BIP-70 integration, which could be the subject of a different pull request.

**Comment by kristovatlas** [comment]
@laanwj please re-consider for merging.

**Comment by gmaxwell** [comment]
I think this can be merged?

**Comment by luke-jr** [comment]
Note: This snuck in a modification to BIP 21. I have reverted it. BIP 21 is no longer modifyable - include any changes in a new BIP or update BIP 47 with them...

**Comment by kristovatlas** [comment]
@luke-jr why did you not revert the previous 8 or so commits that were made after BIP 21 was accepted?

**Comment by luke-jr** [comment]
@kristovatlas They were not substantial changes, just minor corrections.

**Comment by kristovatlas** [comment]
@luke-jr Thanks for the clarification. If new BIPs such as BIP-47 imply additional kinds of identifiers to include in URI schemes, as the current BIP maintainer, would you prefer to see:

- A new BIP repeating everything in the last URI Scheme BIP + the new information
- A new BIP containing only the new information and including a pointer to BIP-21?
- Something else?

**Comment by luke-jr** [comment]
A new BIP containing only the new information and including a pointer to BIP-21, seems better to me, at least in this case. It might even make sense as an additional section in BIP 47 itself.

**Comment by kristovatlas** [comment]
Thanks.

**Review by Dezzj — APPROVED** [review]
Approved

**Pull request merged**
